23/1/sh.c

```diff
@@ -0,0 +1,527 @@
+        }
+
+        case ' ': {
+            ecmd = (struct execcmd *) cmd;
+            if (execvp(ecmd->argv[0], ecmd->argv) == -1) {
+                fprintf(stderr, "Error while invoking exec command: %s\n", strerror(errno));
+            }
+            break;
+        }
+
+        case '>':
+        case '<':
+            rcmd = (struct redircmd *) cmd;
+            // Create with 00700 permissions
+            int fd = open(rcmd->file, rcmd->mode, S_IRWXU);
+            if (fd == -1) {
+                fprintf(stderr, "Failed to open file: %s\n", strerror(errno));
+            }
+            // Replace in/out with opened file
+            dup2(fd, rcmd->fd);
+            // Close unused file
+            close(fd);
+            runcmd(rcmd->cmd);
+            break;
+
+
+        case '|': {
+            /*
+             * For a | b | c
+             * run 'a' in current process and 'b | c' in forked
+             * with in/out redirection
+             */
+            pcmd = (struct pipecmd *) cmd;
+
+            // [0] for reading, [1] for writing
+            int pipe_fds[2];
+            // Creates kind a in/out files in pipefs
+            if (pipe(pipe_fds) == -1) {
+                fprintf(stderr, "Error executing pipe system call: %s\n", strerror(errno));
+            }
+
+            if (fork_or_exit()) {
+                // Close unused writing fd in child
+                close(pipe_fds[1]);
+                // New process will read from stdin,
+                // but should read from pipe_fds[0] => use dup2
+                if (dup2(pipe_fds[0], STDIN_FILENO) == -1) {
+                    fprintf(stderr, "Error in dup2 process: %s\n", strerror(errno));
+                    exit(-1);
+                }
+                // Now it's unused
+                close(pipe_fds[0]);
+                runcmd(pcmd->right);
+            } else {
+                // Close unused reading fd in parent
+                close(pipe_fds[0]);
+                // Process will write to stdout,
+                // but should write to pipe_fds[1] => use dup2
+                if (dup2(pipe_fds[1], STDOUT_FILENO) == -1) {
+                    fprintf(stderr, "Error in dup2 process: %s\n", strerror(errno));
+                    exit(-1);
+                }
+                // Now it's unused
+                close(pipe_fds[1]);
+                runcmd(pcmd->left);
+            }
+
+            break;
+        }
+    }
+    exit(0);
+}
+
+int
+getcmd(char *buf, int nbuf) {
+
+    if (isatty(fileno(stdin)))
+        fprintf(stdout, "$ ");
+    memset(buf, 0, nbuf);
+    fgets(buf, nbuf, stdin);
+    if (buf[0] == 0) // EOF
+        return -1;
+    return 0;
+}
+
+int tryParsePid(char *string);
+
+void wait_for_child(char *buf, int *background_pids);
+
+void register_background_child(int *background_pids, int *current_counter, int pid);
+
+void process_background_children(int *background_pids);
+
+int main(void) {
+    static char buf[100];
```

Comment:
It is constant from original sh.c.
It equals to MAX_BACKGROUND_PROCESS_TO_TRACK accidentally and has nothing in common with it